enhancements/node-artifacts/certificates-and-pull-secrets.md

```diff
@@ -0,0 +1,199 @@
+---
+title: node-artifacts-during-build-and-image-stream-import
+authors:
+  - "@rmarasch"
+reviewers:
+  - "@dmage"
+  - "@bparees"
+  - "@adambkaplan"
+approvers:
+  - "@dmage"
+  - "@bparees"
+  - "@adambkaplan"
+creation-date: 2019-12-02
+last-updated: 2019-12-03
+status: provisional
+---
+
+# Using node certificates/pull secrets during build and image stream import
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Allow OpenShift users to import and use images from any registry configured
+during or after the cluster installation by sharing node's artifacts(certificates
+and pull secrets) with `openshift-api` and `builder` pods.
```

Comment:
I don't think this proposal should worry about sharing the node's certificates.  We have sufficient mechanisms for providing CAs that all builds/imagestreams can trust at a global cluster level.

We're only missing the creds aspect.

Yes having builds+imagestreams use the node's CAs would remove one more possible delta between the experience of a pod pulling and image and an imagestream/build doing so, but i don't think it's a sufficient pain point to justify the additional complexity of adding yet another CA source to the transport CA loading hierarchy.

Reply:
That would hugely simplify things. What do you reckon @adambkaplan ? I remember this decision of incorporating CAs to this enhancements came out of one of our team planning sessions.

Reply:
@bparees there are also client certificates (we don't support them yet), mirroring configuration (couple of different implementations). Can we add such things to a proposal as a vision how that can be implemented if we need them in the future? CAs is kind of a retrospective thing, we don't need to reimplement them if they are already implemented, but if we find another component that lack support of CA (unlikely), they can benefit from this approach as well.

Reply:
@dmage this enhancement is specifically about ensuring that components that talk to registries (imagestreams, builds) have the same experience/behavior as nodes/pods do (meaning if a pod can pull the image, an imagestream+build can also access it).


Mirroring config we've already covered in imagestreams+builds, right (we use the same config as the node)?

Do we currently allow client certificates to be placed on nodes to perform image pulling?  Even if we do, i think i'd prefer to see that in a separate proposal or a future iteration on this proposal, i'd like to keep this particular item scoped to the smallest thing that's going to make our users lives much easier and i think that is just using the node creds for auth in imagestreams+builds.

And if you're just talking about allowing the use of client certs for auth independent of what nodes do, i think that's definitely a separate RFE+enhancement proposal.

Reply:
@bparees I'm exactly about having the same experience/behavior as nodes/pods do. So far mirroring and CAs give slightly different experience for pods, openshift-apiserver, builds, and image-registry pull-through.

Mirroring config we've covered, but we also have a desire to use `containers/image` in openshift-apiserver and image-registry. This package needs configuration to be provided through `/etc/containers/...`. And I'd prefer to have this configuration from the node rather than implement another converter from k8s objects to a config file.

I'm not aware of client certificates in OCP 4.x, so it was just an example of the artifact that nodes can already use, but other implementations are not aware of.

Reply:
honestly my preference would be to tie ourselves as little as possible to the node configuration and have k8s objects be the source of truth for both the nodes + other components.

but in the interest of expediency, and because nodes can have configuration that's not easily read from k8s objects (because the MCO allows users to inject arbitrary files on the node), mounting/consuming the node files is easier and i'm ok with that in some cases (such as here).  I don't think we should treat as the default/preferred approach though.

Reply:
We do have [this](https://github.com/openshift/enhancements/pull/115) other proposal on CAs that we need to keep sync with as we may have some overlapping areas with this one.

Reply:
After team discussions during Arch meeting I moved "certificates" out of this enhancement proposal's scope.

Reply:
@ricardomaraschini please add the result of our discussions to the "Non-Goals" section of the proposal.

Reply:
@adambkaplan nice catch. I have explicitly added CAs and "mirroring config" to the non-goal section.

Reply:
We removed the certificates our of this proposal. Resolving this one.